src/storage/src/hummock/cache/block_cache.rs

```diff
@@ -0,0 +1,326 @@
+// Copyright 2022 Singularity Data
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::hash_map::DefaultHasher;
+use std::hash::{Hash, Hasher};
+use std::ops::Deref;
+use std::sync::Arc;
+
+use bytes::{Buf, BufMut, Bytes};
+use futures::Future;
+use risingwave_common::cache::{CachableEntry, LruCache, LruCacheEventListener};
+use risingwave_common::config::FileCacheConfig;
+use risingwave_hummock_sdk::HummockSstableId;
+
+use crate::hummock::{
+    Block, HummockError, HummockResult, TieredCache, TieredCacheKey, TieredCacheOptions,
+};
+
+const MIN_BUFFER_SIZE_PER_SHARD: usize = 32 * 1024 * 1024;
+
+enum BlockEntry {
+    Cache(CachableEntry<(HummockSstableId, u64), Box<Block>>),
+    Owned(Box<Block>),
+    RefEntry(Arc<Block>),
+}
+
+pub struct BlockHolder {
+    _handle: BlockEntry,
+    block: *const Block,
+}
+
+impl BlockHolder {
+    pub fn from_ref_block(block: Arc<Block>) -> Self {
+        let ptr = block.as_ref() as *const _;
+        Self {
+            _handle: BlockEntry::RefEntry(block),
+            block: ptr,
+        }
+    }
+
+    pub fn from_owned_block(block: Box<Block>) -> Self {
+        let ptr = block.as_ref() as *const _;
+        Self {
+            _handle: BlockEntry::Owned(block),
+            block: ptr,
+        }
+    }
+
+    pub fn from_cached_block(entry: CachableEntry<(HummockSstableId, u64), Box<Block>>) -> Self {
+        let ptr = entry.value().as_ref() as *const _;
+        Self {
+            _handle: BlockEntry::Cache(entry),
+            block: ptr,
+        }
+    }
+}
+
+impl Deref for BlockHolder {
+    type Target = Block;
+
+    fn deref(&self) -> &Self::Target {
+        unsafe { &(*self.block) }
+    }
+}
+
+unsafe impl Send for BlockHolder {}
+unsafe impl Sync for BlockHolder {}
+
+#[derive(PartialEq, Eq, Hash, Clone, Debug)]
+pub struct CacheKey {
+    sst: u64,
```

Comment:
May rename `sst` to `sst_id` and `idx` to `block_idx`.

Besides, the `TODO` seems unnecessary to me since even if we use u32, we still need to add 4 bytes padding to ensure alignment.